futures-util/src/lock.rs

```diff
@@ -135,27 +135,30 @@ impl<T> BiLock<T> {
     /// Note that the returned future will never resolve to an error.
     pub fn lock(&self) -> BiLockAcquire<T> {
         BiLockAcquire {
-            inner: self,
+            bilock: self,
         }
     }
 
     /// Attempts to put the two "halves" of a `BiLock<T>` back together and
     /// recover the original value. Succeeds only if the two `BiLock<T>`s
     /// originated from the same call to `BiLock::new`.
-    pub fn reunite(self, other: Self) -> Result<T, ReuniteError<T>> {
-        if &*self.inner as *const _ == &*other.inner as *const _ {
+    pub fn reunite(self, other: Self) -> Result<T, ReuniteError<T>>
+    where
+        T: Unpin,
```

Comment:
I think the `Unpin` bound is not required because it takes self by value, so it couldn't have been pinned before.

Reply:
The bilock couldn't have been pinned, but this PR changes to assume that the innards of a bilock (behind the arc) are always pinned. This is required in order to call `self: PinMut<Self>` methods on the innards of the bilock.

Reply:
Ah, I see